### PR TITLE
Tag api endpoints

### PR DESCRIPTION
**Pull request by replay**
Add the API endpoints for:

`/tags` (with optional `filter`)
`/tags/<tag>` (with optional `filter`)
`/tags/findSeries` (with list of `expr`)

The goal is that the API is exactly the same like documented in the graphite docs: http://graphite.readthedocs.io/en/latest/tags.html#querying
The `pretty` parameter has been ignored for now.

I think that once we change the main index to use the numeric `idx.MetricID` type instead of `string`s then the performance fo calls like `/tags/<tag>` might improve, because they rely heavily on ID lookups in the main index.

### CODE REVIEW

File: api/routes.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ func (s *Server) RegisterRoutes() {
 	r.Combo("/index/list", ready, bind(models.IndexList{})).Get(s.indexList).Post(s.indexList)
 	r.Combo("/index/delete", ready, bind(models.IndexDelete{})).Get(s.indexDelete).Post(s.indexDelete)
 	r.Combo("/index/get", ready, bind(models.IndexGet{})).Get(s.indexGet).Post(s.indexGet)
+	r.Combo("/index/tags", ready, bind(models.IndexTags{})).Get(s.indexTags).Post(s.indexTags)
+	r.Combo("/index/tags/findSeries", ready, bind(models.IndexTagFindSeries{})).Get(s.indexTagFindSeries).Post(s.indexTagFindSeries)
+	r.Combo("/index/tags/:tag([0-9a-zA-Z]+)", ready, bind(models.IndexTagDetails{})).Get(s.indexTagDetails).Post(s.indexTagDetails)
```
Comment (Dieterbe) [inline api/routes.go]: does it make sense to have a subhierachy here? 
this is meant to be a 1:1 mapping of http url to index operation, also naming wise, so I think this would make more sense:
```
"/index/tags" -> s.indexTags
"/index/find_by_tag"-> s.indexFindByTag
"/index/tag_details" -> s.indexTagDetails // specify the tag via the model, not via the url
```

File: api/graphite.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ import (
 	tags "github.com/opentracing/opentracing-go/ext"
 	"github.com/raintank/dur"
 	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/tinylib/msgp/msgp"
```
Comment (Dieterbe) [inline api/graphite.go]: a bit weird that this library now needs to be aware of the msgp serialization whereas in other codepaths this is handled by the callers, can we clean this up a bit?

File: api/models/graphite.go [hunk 3]
```diff
@@ -52,6 +52,42 @@ func (gr GraphiteRender) Validate(ctx *macaron.Context, errs binding.Errors) bin
...
+	Filter string `json:"filter"`
+	From   int64  `json:"from"`
+}
+
+type GraphiteTagsResp []GraphiteTagResp
+
+type GraphiteTagResp struct {
+	Tag string `json:"tag"`
+}
+
```
Comment (Dieterbe) [inline api/models/graphite.go]: the modeling here seems a bit weird. a named structure with 1 string member? 
can we do something like `type GraphiteTagsResp []string` or do it like you did `GraphiteTagFindSeriesResp` ?

File: api/models/graphite.go [hunk 4]
```diff
@@ -52,6 +52,42 @@ func (gr GraphiteRender) Validate(ctx *macaron.Context, errs binding.Errors) bin
...
+type GraphiteTagDetailsValueResp struct {
+	Count uint64 `json:"count"`
+	Value string `json:"value"`
+}
+
+type GraphiteTagDetailsResp struct {
+	Tag    string                        `json:"tag"`
+	Values []GraphiteTagDetailsValueResp `json:"values"`
```
Comment (Dieterbe) [inline api/models/graphite.go]: can we stick to the convention of first declaring a thing, and then the more specific thing that it relies on?
e.g. first `GraphiteTagDetailsResp`, then `GraphiteTagDetailsValueResp`
I think this is a useful convention, and we also do it in many places for functions and then the sub-functions that they call.

File: idx/memory/memory.go [hunk 5]
```diff
@@ -376,76 +376,161 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
...
 	}
 
 	m.RLock()
 	defer m.RUnlock()
 
-	var tags TagIndex
-	var ok bool
-	if tags, ok = m.Tags[orgId]; !ok {
-		return nil
+	tree, ok := m.tags[orgId]
```
Comment (Dieterbe) [inline idx/memory/memory.go]: because we have:
```
type MemoryIdx struct {
	sync.RWMutex
	DefById map[string]*idx.Archive
	Tree    map[int]*Tree
	tags    map[int]TagIndex
}
```
let's stick to the name tag (or tagIndex) when we get a substructure from tags, and the name tree when we get a substructure from Tree

File: api/routes.go (hunk 1 above)
Comment (replay) [reply api/routes.go]: When I named them I basically took the urls as described in the graphite docs and pre-pended `/index` because those are all the internal calls to only be used by other cluster: nodes.http://graphite.readthedocs.io/en/latest/tags.html#exploring-tags
I'd actually like the urls you suggest more as well, but I think that would be unnecessarily confusing or not?

File: api/routes.go [hunk 6]
```diff
@@ -49,5 +52,7 @@ func (s *Server) RegisterRoutes() {
 	r.Combo("/metrics/find", withOrg, ready, bind(models.GraphiteFind{})).Get(s.metricsFind).Post(s.metricsFind)
 	r.Get("/metrics/index.json", withOrg, ready, s.metricsIndex)
 	r.Post("/metrics/delete", withOrg, ready, bind(models.MetricsDelete{}), s.metricsDelete)
-
+	r.Combo("/tags", withOrg, ready, bind(models.GraphiteTags{})).Get(s.graphiteTags).Post(s.graphiteTags)
```
[the remark this answers is not in the document]
Comment (replay) [reply api/routes.go]: True, should move it to `/metrics`.
Isn't the local index part of the cluster? My idea was to have a method that's responsible for dealing with http stuff, like parsing request and building response etc, and another one to implement the logic. but yeah, will make it conform with the other parts.

[the remark this answers is not in the document]
Comment (replay) [reply api/routes.go]: how about i move `clusterQuery` to `/api/cluster.go`? that one already has `msgp` imported anyway

File: api/graphite.go (hunk 2 above)
Comment (replay) [reply api/graphite.go]: how about i move clusterQuery to /api/cluster.go? that one already has msgp imported anyway

Comment (replay) [reply api/graphite.go]: like https://github.com/grafana/metrictank/commit/6924ec6f5305aa07839303a30080816e4f1a148b

File: api/models/graphite.go (hunk 3 above)
Comment (replay) [reply api/models/graphite.go]: hmmm... then how could i end up with the same format as graphite outputs?
the output needs to be of this format:
```
[
  {
    "tag": "name"
  },
  {
    "tag": "datacenter"
  },
  {
    "tag": "rack"
  },
  {
    "tag": "server"
  }
]```

File: api/models/graphite.go (hunk 4 above)
Comment (replay) [reply api/models/graphite.go]: :+1: done

File: idx/memory/memory.go (hunk 5 above)
Comment (replay) [reply idx/memory/memory.go]: done https://github.com/grafana/metrictank/pull/750/commits/5118967a09b4c63b060f82eb487ad2d335eb8e74

File: api/models/graphite.go (hunk 4 above)
Comment (replay) [reply api/models/graphite.go]: https://github.com/grafana/metrictank/pull/750/commits/c3cbba2576a3016092f959a5bb224fec47ca2507

File: api/routes.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (replay) [reply api/routes.go]: actually if you think the utility function `clusterQuery` makes sense the way it is we could probably use it on other endpoint handlers that trigger cross-cluster calls too, then they would all be conform. 
in the beginning i had separate cross-cluster calling functions for each of the endpoints similar to `findSeriesRemote`, but it bothered me that there's so much duplication in how those were structured, that's why i ended up merging all of the duplication into the utility function. 
i didn't modify any of the older functions yet to use that too, but could do that.

[the remark this answers is not in the document]
Comment (replay) [reply api/routes.go]: maybe i should rename it to `peerQuery` though. because it only queries the cluster peers, but excludes the local index.

File: api/models/graphite.go (hunk 3 above)
Comment (Dieterbe) [reply api/models/graphite.go]: oh well if that's a requirement, not much we can I do think.
Is there a reason the output format can't just be like this? do we expect to add more fields for each tag later?
```
[ "name", "datacenter", "rack", "server"]
```

Comment (DanCech) [reply api/models/graphite.go]: yes, it's designed to be extensible

File: api/models/graphite.go (hunk 4 above)
Comment (Dieterbe) [reply api/models/graphite.go]: i don't have strong opinions on this and we can research this topic a bit more to see what the best practices are. see #751 but yeah for now let's stick to this approach just for consistency.

File: api/routes.go (hunk 1 above)
Comment (Dieterbe) [reply api/routes.go]: personally I like the 1:1 mapping of http path to index operation, and the more simple we can keep our patterns and naming of functions etc, the easier this becomes to read IMHO.

**Comment by Dieterbe** [comment]
i just added a commit to your branch. i always disliked that the method docs for the index were so separate from their signatures, making things hard to tie together, so i cleaned that up. this is how it should be done per godoc convention anyway.

**Comment by replay** [comment]
@shanson7 i seem to be unable to add you as a reviewer. But you might also be interested in this PR

File: api/routes.go (hunk 1 above)
Comment (DanCech) [reply api/routes.go]: This needs to match the graphite url hierarchy so that it is queryable by the graphite http tagdb

Comment (replay) [reply api/routes.go]: @DanCech actually these urls are only used for communication between cluster peers, to fanout a query that came in. But the reason why I named them the way they are now is because that makes it easy to figure out which of those internal endpoints is used by which of the graphite endpoints.

File: api/cluster.go [hunk 7]
```diff
@@ -106,6 +108,37 @@ func (s *Server) indexFind(ctx *middleware.Context, req models.IndexFind) {
...
+func (s *Server) indexTagFindSeries(ctx *middleware.Context, req models.IndexTagFindSeries) {
+	metrics, err := s.MetricIndex.FindByTag(req.OrgId, req.Expressions, req.From)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, err.Error()))
+		return
+	}
+	metricNames := make([]string, 0, len(metrics))
+	for _, metric := range metrics {
+		metricNames = append(metricNames, metric)
+	}
```
Comment (Dieterbe) [inline api/cluster.go]: why the copy here?
it looks like this slice is generated from scratch in `(m *MemoryIdx) resolveIDs` so we can just reuse it safely.

Comment (Dieterbe) [reply api/cluster.go]: in general, I think we should also be a bit more precise around semantics of reuse/concurrent access. and whenever returning types that are/have pointers or references (pointers, slices, structs with slices), clarify what the caller is allowed or supposed to do.
for example for MetricIndex.FindByTag which returns a `[]string`. This concern shouldn't leak across api boundaries, i.e. we shouldn't have to go digging into index implementations to figure out if we need to copy or not.

I think we can also take inspiration from various packages (both in standard libraries and 3rd packages) and take this a step further. e.g.:
- https://golang.org/pkg/io/#Reader
- https://github.com/tinylib/msgp/blob/b2b6a672cf1e5b90748f79b8b81fc8c5cf0571a1/msgp/write.go#L90
We would have the caller pass in a `[]string`, which is then used to populate the response and then returned like in the msgp example. the benefit is then that the ownership of that slice is more obviously the caller's and the caller can also control allocations by using a pool. (the MetricIndex can't work with a pool directly because when it returns data it doesn't know when the data is no longer used. unless we add an extra callback function but that just complicates things)
if the caller doesn't care about any of this (and initially we often don't, until we profile and demonstrate we really need a pool or something) it can just pass `nil` as input arg.

so the implication here is that the first arg of `FindByTag` would be a `[]string` and we change the method documentation to say something like `	It stores the metric names that match the given conditions into the provided slice and returns the extended slice and any errors encountered.`

let me know if any of this doesn't make sense.

File: idx/memory/memory.go [hunk 8]
```diff
@@ -376,76 +376,161 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
...
+
+		res = append(res, key)
+	}
+
+	return res, nil
+}
+
+// resolves a list of ids (TagIDs) into a list of metric names
+// assumes that at least a read lock is already held by the caller
+func (m *MemoryIdx) resolveIDs(ids TagIDs) []string {
```
Comment (Dieterbe) [inline idx/memory/memory.go]: godoc format `resolveIDs resolves...`

File: idx/memory/memory.go [hunk 9]
```diff
@@ -376,76 +376,161 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
...
+	res := make([]string, len(ids))
+	i := uint32(0)
+	for id := range ids {
+		def, ok := m.DefById[id.String()]
+		if !ok {
+			continue
```
Comment (Dieterbe) [inline idx/memory/memory.go]: this looks like an index corruption case we should log and increment metric for

File: idx/memory/memory.go [hunk 10]
```diff
@@ -376,76 +376,161 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
...
+	tags, ok := m.tags[orgId]
+	if !ok {
+		return nil, nil
 	}
 
-	result := make(map[string]uint32)
+	values, ok := tags[key]
+	if !ok {
+		return nil, nil
+	}
```
Comment (Dieterbe) [inline idx/memory/memory.go]: small, but well spotted optimization :)

File: idx/memory/memory.go [hunk 11]
```diff
@@ -376,76 +376,161 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
...
+
+				if def.LastUpdate < from {
+					continue
+				}
 
-			valueCnt++
+				count++
+			}
+		} else {
+			count += uint64(len(ids))
```
Comment (Dieterbe) [inline idx/memory/memory.go]: also well spotted :+1:

File: idx/memory/memory.go [hunk 12]
```diff
@@ -376,76 +376,161 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
 	return archives
 }
 
-func (m *MemoryIdx) Tag(orgId int, tag string, from int64) map[string]uint32 {
+func (m *MemoryIdx) TagDetails(orgId int, key, filter string, from int64) (map[string]uint64, error) {
```
Comment (Dieterbe) [inline idx/memory/memory.go]: the change from uint32 to uint64 is only useful if we expect any tagkey-tagval combo to occur more than 4 billion times right? fine by me, hope we hit that number sometime ;) important to note here that these counts are per-instance, but they get added up with all other instances, where we get larger aggregate counts.

File: api/routes.go (hunk 1 above)
Comment (Dieterbe) [reply api/routes.go]: > the reason why I named them the way they are now is because that makes it easy to figure out which of those internal endpoints is used by which of the graphite endpoints.

this won't work well anymore I think when these endpoints start to get used for other things. for example in the future MT will start hitting some of these endpoints as it is responding to render requests. that's why i think it's more future proof that an endpoint http path represents the operation that will be performed. also for diagnosing issues i think it makes morse sense.

File: api/routes.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dieterbe) [reply api/routes.go]: is it going to be a problem for graphite to have them in MT under `/metrics` ? 
also, as mentioned in my other comment for the `/index/tags/:tag([0-9a-zA-Z]+)` route, do we need to put the tag in the http path or can we just pass it as a field?
currently all of our other routes pass everything as fields, we don't do fancy stuff like this and I think that's good. for one reason, we mark operation names in our traces by the http path, and those can't have too high cardinality. whereas field names can have pretty much unbounded cardinality.

File: api/cluster.go [hunk 13]
```diff
@@ -153,3 +186,69 @@ func (s *Server) indexDelete(ctx *middleware.Context, req models.IndexDelete) {
...
+// peerQuery takes a request and the path to request it on, then fans it out
+// across the cluster, except to the local peer.
+// ctx:          request context
+// data:         request to be submitted
+// name:         name to be used in logging & tracing
+// path:         path to request on
+// respTemplate: the response object into which the responses shall
+//               be deserialized. each sub-request will get a copy of this
+//               template and the response will be unarshalled into that copy.
+//               (generic without generics)
```
Comment (Dieterbe) [inline api/cluster.go]: this seems incorrect.
doing `resp := respTemplate` is not al copy. see https://play.golang.org/p/vILbXj3BfU for a demo.
so seems to me like this will corrupt the responses of different nodes, effectively what would happen I think is that all the responses are the response of the last node to respond.

Comment (Dieterbe) [reply api/cluster.go]: (the demo is based on the code for Server.clusterTagDetails )

File: api/graphite.go [hunk 14]
```diff
@@ -673,3 +673,143 @@ func getLocation(desc string) (*time.Location, error) {
 	}
 	return time.LoadLocation(desc)
 }
+
+func (s *Server) graphiteTagDetails(ctx *middleware.Context, request models.GraphiteTagDetails) {
+	tag := ctx.Params(":tag")
+	if len(tag) <= 0 {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, "not tag specified"))
```
Comment (Dieterbe) [inline api/graphite.go]: another benefit of specifying tag as a field is that we get validation and don't need to validate in our handlers anymore.

File: api/models/node.go [hunk 15]
```diff
@@ -35,6 +35,53 @@ func (i IndexList) Trace(span opentracing.Span) {
 func (i IndexList) TraceDebug(span opentracing.Span) {
 }
 
+type IndexTagFindSeries struct {
+	OrgId       int      `json:"orgId" binding:"Required"`
+	Expressions []string `json:"expressions"`
```
Comment (Dieterbe) [inline api/models/node.go]: I noticed in `models.GraphiteTagFindSeries{}` this filed is called `Expr`. let's name this field the same way?

File: api/models/cluster.go [hunk 16]
```diff
@@ -24,3 +24,18 @@ type GetDataResp struct {
...
+}
+
+//go:generate msgp
+type IndexTagDetailsResp struct {
+	Values map[string]uint64 `json:"values"`
+}
+
+//go:generate msgp
+type IndexTagFindSeriesResp struct {
+	Series []string `json:"series"`
```
Comment (Dieterbe) [inline api/models/cluster.go]: the docs for FindByTag talk about "metric names", when we call it we do "metrics, err := s.MetricIndex.FindByTag"
and here we talk about series. can we call this "metrics" here too?

File: api/graphite.go [hunk 17]
```diff
@@ -673,3 +673,143 @@ func getLocation(desc string) (*time.Location, error) {
...
+	for _, resp := range responses {
+		for _, series := range resp.(*models.IndexTagFindSeriesResp).Series {
+			seriesSet[series] = struct{}{}
+		}
+	}
+
+	series := make([]string, 0, len(seriesSet))
+	for s := range seriesSet {
+		series = append(series, s)
+	}
```
Comment (Dieterbe) [inline api/graphite.go]: For the record, I assume the intermediate set is to remove duplicates.
1) we may get the same metric from multiple instances
2) we may get the same metric from the same instance 

I think both of these could happen when people start publishing the same metric but under a different interval, and/or it goes to a different partition (FYI)

note that we do not have duplicate surpression in FindByTag. I think we should add it there too.

File: idx/memory/tag_query_test.go [hunk 18]
```diff
@@ -286,11 +260,9 @@ func TestGetByTag(t *testing.T) {
 		if len(res) != len(tc.expectation) {
 			t.Fatalf("Result does not match expectation for expressions %+v\nGot:\n%+v\nExpected:\n%+v\n", tc.expressions, res, tc.expectation)
```
Comment (Dieterbe) [inline idx/memory/tag_query_test.go]: please expected first, got last

File: idx/memory/tag_query_test.go [hunk 19]
```diff
@@ -182,33 +183,6 @@ func TestTagExpressionQueryByTagWithFrom(t *testing.T) {
...
-	if res["value1"] != 2 {
-		t.Fatalf("Expected %d results, but got %d", 4, len(res))
-	}
-
-	res = memIdx.Tag(1, "key1", 4)
-	if res["value1"] != 1 {
-		t.Fatalf("Expected %d results, but got %d", 4, len(res))
-	}
-}
-
```
Comment (Dieterbe) [inline idx/memory/tag_query_test.go]: why is this removed?

File: api/routes.go (hunk 1 above)
Comment (replay) [reply api/routes.go]: ok, renaming

Comment (replay) [reply api/routes.go]: https://github.com/grafana/metrictank/pull/750/commits/d30bbe7655357a99325da08b607d45e088b0ce28

File: api/routes.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (replay) [reply api/routes.go]: While looking at it again i realized that in the `/index/tags/<tag>` route that `tag` from the url wasn't even used, so to answer your question, no we can skip that: https://github.com/grafana/metrictank/commit/d30bbe7655357a99325da08b607d45e088b0ce28

File: api/cluster.go (hunk 7 above)
Comment (replay) [reply api/cluster.go]: Regarding passing the slice to store results in as an argument:
- I understand the benefit of that "implied" ownership/who's responsible for it

But I think apart from that there are more disadvantages than advantages:
- F.e. `resolvIDs` will preallocate the required number of elements in the result slice because it already knows them at the time the slice is initialized
- Even if the documentation says that the caller just needs to pass the slice to store results in, I think probably almost everybody who's going to use that will still look at the code of that called function anyway. 
- It just makes the function signature longer

File: idx/memory/memory.go (hunk 12 above)
Comment (replay) [reply idx/memory/memory.go]: yes, better be future proof :) It also has the advantage that in `clusterTagDetails` no new map needs to be instantiated for the sums of all instance counts across the cluster, and I'd guess that when aggregating across a large cluster the 4 billion count is easier to reach than in a single MT.

File: api/cluster.go (hunk 13 above)
Comment (replay) [reply api/cluster.go]: yeah true. i guess that way of making it generic, by passing in a template to copy, is kind of ugly anyway. what would be better is to first get the list of peers in the caller, generate the slice of responses based on the number of peers, and then pass both slices into `peerQuery()`

Comment (replay) [reply api/cluster.go]: that above suggestion doesn't work because `[]<specific type>` will not be accepted if the signature expects `[]interface`: https://github.com/golang/go/wiki/InterfaceSlice
but what do you think about that: https://github.com/grafana/metrictank/pull/750/commits/18b2257206765b49ec25c0a30a93e91f2332e4d2

File: api/graphite.go (hunk 14 above)
Comment (replay) [reply api/graphite.go]: for this one we still need to do that, because that's the url schema that graphite expects. but yes, for the cluster-internal call on `/index` we can omit that

File: api/models/node.go (hunk 15 above)
Comment (replay) [reply api/models/node.go]: done :+1:

File: api/models/cluster.go (hunk 16 above)
Comment (replay) [reply api/models/cluster.go]: done :+1:

File: api/graphite.go (hunk 17 above)
Comment (replay) [reply api/graphite.go]: What's the benefit of adding it to `FindByTag`? currently there is no way to call `FindByTag` without also going through the deduplication in `clusterTagFindSeries`. I guess one could argue that in the future such a case might get added, but until then this would just be a waste of cpu cycles or not?

File: idx/memory/tag_query_test.go (hunk 19 above)
Comment (replay) [reply idx/memory/tag_query_test.go]: because it is redundant with the tests in `memory_find_test.go`, such as `TestTagDetailsWithFrom()`

File: api/cluster.go (hunk 13 above)
Comment (Dieterbe) [reply api/cluster.go]: > because []<specific type> will not be accepted if the signature expects []interface

yeah bummer.  i've also run into this before.

> but what do you think about that: 18b2257

it's unfortunate that we have to move half of peerQueryi's processing logic back into the callers. that starts to defeat the point of the reusable function. i wonder if the actual type can be passed in and then peerQuery could instantiate however many of the type it needs via https://golang.org/pkg/reflect/#New

**Comment by replay** [comment]
@Dieterbe i think all the comments have either been replied to or the requests/suggestions have been added

File: api/graphite.go (hunk 17 above)
Comment (Dieterbe) [reply api/graphite.go]: you're right. so until then, let's add a warning to the function (both implementation and index method) that it may return duplicates

File: api/cluster.go (hunk 7 above)
Comment (Dieterbe) [reply api/cluster.go]: > F.e. resolvIDs will preallocate the required number of elements in the result slice because it already knows them at the time the slice is initialized

you can still do the equivalent with a slice that is passed in (whether it is nil or has any length already) see https://github.com/raintank/schema/blob/4a55004ab1f55549c476850030990da9f72574b5/metric.go#L72 however this is non-trivial, so i'm fine with leaving it like this for now (we can re-discuss when we need to introduce a pool or something)
but my point (I think) remains that these semantics of what the caller can expect around reuse of the slice, should be documented in the method docs. 
e.g. they can just say something like "the caller can do to the returned slice what it wants".
or should this be implied and only ever document cases where this is *not* the case?

File: idx/memory/tag_query_test.go (hunk 18 above)
Comment (Dieterbe) [reply idx/memory/tag_query_test.go]: like last time, i was commenting on the Fatalf line ;-)

File: api/routes.go (hunk 1 above)
Comment (Dieterbe) [reply api/routes.go]: why is all the stuff in between the `find_by_tag` route and the `s.MetricIndex.FindByTag()` call, still called `indexTagFindSeries` ? (both the function, and input and output models)

File: api/models/node.go [hunk 20]
```diff
@@ -35,6 +35,53 @@ func (i IndexList) Trace(span opentracing.Span) {
...
+	span.SetTag("from", t.From)
+}
+
+func (i IndexTagFindSeries) TraceDebug(span opentracing.Span) {
+}
+
+type IndexTagDetails struct {
+	OrgId  int    `json:"orgId" binding:"Required"`
+	Filter string `json:"filter"`
+	Tag    string `json:"tag"`
```
Comment (Dieterbe) [inline api/models/node.go]: are empty tag keys legal? otherwise add `binding:"Required"` here

File: api/cluster.go (hunk 13 above)
Comment (replay) [reply api/cluster.go]: i think that would be nice. but without knowing it, just based on what i understand the `reflect` lib does, i'd guess that's incurring overhead

File: api/graphite.go (hunk 17 above)
Comment (replay) [reply api/graphite.go]: https://github.com/grafana/metrictank/pull/750/commits/2d9cf9ef11fe59d2ac0b90d10811be5d9c9dad03

File: idx/memory/tag_query_test.go (hunk 18 above)
Comment (replay) [reply idx/memory/tag_query_test.go]: lol, right, sorry https://github.com/grafana/metrictank/pull/750/commits/5deda8aea2603aa694abc9009aeb26b881bcddaf

File: api/models/node.go (hunk 20 above)
Comment (replay) [reply api/models/node.go]: you're right, that should be required https://github.com/grafana/metrictank/pull/750/commits/5deda8aea2603aa694abc9009aeb26b881bcddaf

File: api/cluster.go (hunk 7 above)
Comment (replay) [reply api/cluster.go]: i think if somebody wants to know what the caller can do to the returned slice they'll simply check the code of the instantiating function. otherwise every single method in the whole code base would need to have comments that explain such things that are usually obvious when looking at the code.

**Comment by replay** [comment]
@Dieterbe before merging i'd like to run that by you to make sure i haven't overseen some edge case in the expression parsing: https://github.com/grafana/metrictank/pull/750/commits/df646b2522ceeecc2bfb579b6475a3ceb1ba02b5

File: api/routes.go (hunk 1 above)
Comment (Dieterbe) [reply api/routes.go]: ping @replay ^^

**Comment by Dieterbe** [comment]
in case you missed it: https://github.com/grafana/metrictank/pull/750#discussion_r146621389
